File: crates/opensymphony-gateway-schema/src/memory_graph.rs
use std::collections::BTreeMap;

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value;

use super::{cursor::StreamCursor, version::SchemaVersion};

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MemoryBundleList {
    pub schema_version: SchemaVersion,
    pub bundles: Vec<MemoryBundleSummary>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MemoryBundleSummary {
    pub id: String,
    pub title: String,
    pub okf_version: String,
    pub visibility: MemoryGraphVisibility,
    pub concept_count: usize,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub updated_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MemoryGraphSnapshot {
    pub schema_version: SchemaVersion,
    pub bundle_id: String,
    pub cursor: StreamCursor,
    pub nodes: Vec<MemoryGraphNode>,
    pub edges: Vec<MemoryGraphEdge>,
    pub communities: Vec<MemoryGraphCommunity>,
    #[serde(default)]
    pub filters_applied: Vec<String>,
    pub generated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MemoryConceptDetail {
    pub schema_version: SchemaVersion,
    pub bundle_id: String,
    pub concept_id: String,
    pub frontmatter_view: MemoryFrontmatterView,
    pub body_markdown: String,
    #[serde(default)]
    pub links: Vec<MemoryGraphLink>,
    #[serde(default)]
    pub citations: Vec<MemoryGraphCitation>,
    #[serde(default)]
    pub source_refs: Vec<MemoryGraphSourceRef>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MemoryCommunityList {
    pub schema_version: SchemaVersion,
    pub bundle_id: String,
    #[serde(default)]
    pub communities: Vec<MemoryGraphCommunity>,
    pub generated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MemorySearchResponse {
    pub schema_version: SchemaVersion,
    pub query: String,
    pub bundle_id: Option<String>,
    pub results: Vec<MemorySearchResult>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MemorySearchResult {
    pub bundle_id: String,
    pub concept_id: String,
    pub title: String,
    pub visibility: MemoryGraphVisibility,
    pub snippet: String,
    #[serde(default)]
    pub areas: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MemoryGraphNode {
    pub id: String,
    pub kind: MemoryGraphNodeKind,
    pub label: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub bundle_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub concept_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub concept_type: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub path_display: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub resource: Option<String>,
    #[serde(default)]
    pub tags: Vec<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub timestamp: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub visibility: Option<MemoryGraphVisibility>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub freshness: Option<MemoryGraphFreshness>,
    #[serde(default)]
    pub warning_count: usize,
    #[serde(default)]
    pub frontmatter_summary: BTreeMap<String, Value>,
    #[serde(default)]
    pub unknown_frontmatter: BTreeMap<String, Value>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub body_preview: Option<String>,
    #[serde(default)]
    pub metrics: MemoryGraphNodeMetrics,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MemoryGraphEdge {
    pub id: String,
    pub kind: MemoryGraphEdgeKind,
    pub source_id: String,
    pub target_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub label: Option<String>,
    #[serde(default)]
    pub unresolved: bool,
    #[serde(default)]
    pub metadata: BTreeMap<String, Value>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MemoryGraphCommunity {
    pub id: String,
    pub label: String,
    #[serde(default)]
    pub node_ids: Vec<String>,
    #[serde(default)]
    pub concept_count: usize,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MemoryFrontmatterView {
    #[serde(default)]
    pub primary: BTreeMap<String, Value>,
    #[serde(default)]
    pub opensymphony: BTreeMap<String, Value>,
    #[serde(default)]
    pub unknown: BTreeMap<String, Value>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct MemoryGraphLink {
    pub target: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub label: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct MemoryGraphCitation {
    pub id: String,
    pub target: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub label: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct MemoryGraphSourceRef {
    pub kind: String,
    pub id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub url: Option<String>,
}

#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
pub struct MemoryGraphNodeMetrics {
    #[serde(default)]
    pub indegree: usize,
    #[serde(default)]
    pub outdegree: usize,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub pagerank: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub community_id: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MemoryGraphUpdatedEvent {
    pub schema_version: SchemaVersion,
    pub bundle_id: String,
    pub cursor: StreamCursor,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MemoryGraphVisibility {
    Public,
    Private,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MemoryGraphFreshness {
    Current,
    Stale,
    Unknown,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MemoryGraphNodeKind {
    Bundle,
    Directory,
    Concept,
    Tag,
    Resource,
    Citation,
    SourceRef,
    Community,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MemoryGraphEdgeKind {
    Contains,
    MarkdownLink,
    ExternalLink,
    Cites,
    TaggedWith,
    DescribesResource,
    ScopedTo,
    SourceSupportedBy,
    SameResource,
}

File: packages/gateway-schema/src/memory_graph.ts
import type { StreamCursor } from "./cursor.js";
import type { SchemaVersion } from "./version.js";

export type MemoryGraphVisibility = "public" | "private";
export type MemoryGraphFreshness = "current" | "stale" | "unknown";

export type MemoryGraphNodeKind =
  | "bundle"
  | "directory"
  | "concept"
  | "tag"
  | "resource"
  | "citation"
  | "source_ref"
  | "community";

export type MemoryGraphEdgeKind =
  | "contains"
  | "markdown_link"
  | "external_link"
  | "cites"
  | "tagged_with"
  | "describes_resource"
  | "scoped_to"
  | "source_supported_by"
  | "same_resource";

export interface MemoryBundleList {
  schema_version: SchemaVersion;
  bundles: MemoryBundleSummary[];
}

export interface MemoryBundleSummary {
  id: string;
  title: string;
  okf_version: string;
  visibility: MemoryGraphVisibility;
  concept_count: number;
  updated_at?: string;
}

export interface MemoryGraphSnapshot {
  schema_version: SchemaVersion;
  bundle_id: string;
  cursor: StreamCursor;
  nodes: MemoryGraphNode[];
  edges: MemoryGraphEdge[];
  communities: MemoryGraphCommunity[];
  metrics?: MemoryGraphSnapshotMetrics;
  filters_applied: string[];
  generated_at: string;
}

export interface MemoryConceptDetail {
  schema_version: SchemaVersion;
  bundle_id: string;
  concept_id: string;
  frontmatter_view: MemoryFrontmatterView;
  body_markdown: string;
  links: MemoryGraphLink[];
  citations: MemoryGraphCitation[];
  source_refs: MemoryGraphSourceRef[];
}

export interface MemoryCommunityList {
  schema_version: SchemaVersion;
  bundle_id: string;
  communities: MemoryGraphCommunity[];
  generated_at: string;
}

export interface MemorySearchResponse {
  schema_version: SchemaVersion;
  query: string;
  bundle_id?: string;
  results: MemorySearchResult[];
}

export interface MemorySearchResult {
  bundle_id: string;
  concept_id: string;
  title: string;
  visibility: MemoryGraphVisibility;
  snippet: string;
  areas: string[];
}

/**
 * One page of completed tasks for the task graph's Completed pane.
 * Primary source is the memory catalog (DuckDB issue capsules with PR
 * evidence); orchestrator-known completed issues not yet captured are
 * merged in with `source: "orchestrator"`.
 */
export interface MemoryCompletedTaskPage {
  schema_version: SchemaVersion;
  bundle_id: string;
  tasks: MemoryCompletedTask[];
  /** Total row count after filtering, before pagination. */
  total: number;
  offset: number;
  limit: number;
  sort: string;
  query?: string;
  generated_at: string;
}

export interface MemoryCompletedTask {
  issue_key: string;
  /** OKF concept id (e.g. `issues/COE-123`); empty for orchestrator rows. */
  concept_id: string;
  bundle_id?: string;
  title: string;
  state?: string;
  milestone?: string;
  url?: string;
  completed_at?: string;
  prs: MemoryTaskPullRequest[];
  source: MemoryCompletedTaskSource;
}

export interface MemoryTaskPullRequest {
  number: number;
  title: string;
  url?: string;
  merged: boolean;
  merged_at?: string;
}

export type MemoryCompletedTaskSource = "memory" | "orchestrator";

export interface MemoryGraphNode {
  id: string;
  kind: MemoryGraphNodeKind;
  label: string;
  bundle_id?: string;
  concept_id?: string;
  concept_type?: string;
  description?: string;
  path_display?: string;
  resource?: string;
  tags: string[];
  timestamp?: string;
  visibility?: MemoryGraphVisibility;
  freshness?: MemoryGraphFreshness;
  warning_count: number;
  frontmatter_summary: Record<string, unknown>;
  unknown_frontmatter: Record<string, unknown>;
  body_preview?: string;
  metrics: MemoryGraphNodeMetrics;
}

export interface MemoryGraphEdge {
  id: string;
  kind: MemoryGraphEdgeKind;
  source_id: string;
  target_id: string;
  label?: string;
  unresolved: boolean;
  metadata: Record<string, unknown>;
}

export interface MemoryGraphCommunity {
  id: string;
  label: string;
  node_ids: string[];
  concept_count: number;
}

export interface MemoryGraphSnapshotMetrics {
  orphan_count: number;
  broken_link_count: number;
  stale_concept_count: number;
  warning_count: number;
}

export interface MemoryFrontmatterView {
  primary: Record<string, unknown>;
  opensymphony: Record<string, unknown>;
  unknown: Record<string, unknown>;
}

export interface MemoryGraphLink {
  target: string;
  label?: string;
}

export interface MemoryGraphCitation {
  id: string;
  target: string;
  label?: string;
}

export interface MemoryGraphSourceRef {
  kind: string;
  id: string;
  url?: string;
}

export interface MemoryGraphNodeMetrics {
  degree?: number;
  indegree: number;
  outdegree: number;
  centrality?: number;
  bridge_score?: number;
  pagerank?: number;
  community_id?: string;
}

export interface MemoryGraphUpdatedEvent {
  schema_version: SchemaVersion;
  bundle_id: string;
  cursor: StreamCursor;
  updated_at: string;
}
